core/coreapi/dag.go

```diff
@@ -56,6 +61,244 @@ func (api *dagAPI) Session(ctx context.Context) ipld.NodeGetter {
+				// 	c,
+				// 	"",
+				// )
+				//
+				// if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
+
+				ret := RootMeta{Cid: c}
+
+				if block, err := api.core.blockstore.Get(c); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if nd, err := ipld.Decode(block); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if err := api.core.pinning.Pin(ctx, nd, true); err != nil {
+					ret.PinErrorMsg = err.Error()
+				} else if err := api.core.pinning.Flush(ctx); err != nil {
+					ret.PinErrorMsg = err.Error()
+				}
+
+				if ret.PinErrorMsg != "" {
+					failedPins++
+				}
+
+				select {
+				case rootsCh <- ret:
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+					return
+				}
+			}
+		}
+	}()
+
+	return rootsCh, errCh
+}
+
+type importResult struct {
+	roots map[cid.Cid]struct{}
+	err   error
+}
+
+func (api *dagAPI) importWorker(ctx context.Context, dir files.Directory, ret chan importResult) {
+	// this is *not* a transaction
+	// it is simply a way to relieve pressure on the blockstore
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(ctx, api)
+
+	roots := make(map[cid.Cid]struct{})
+
+	it := dir.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+
+		// wrap a defer-closer-scope
+		//
+		// every single file in it() is already open before we start
+		// just close here sooner rather than later for neatness
+		// and to surface potential errors writing on closed fifos
+		// this won't/can't help with not running out of handles
+		err := func() error {
+			defer file.Close()
+
+			car, err := gocar.NewCarReader(file)
+			if err != nil {
+				return err
+			}
+
+			// Be explicit here, until the spec is finished
+			if car.Header.Version != 1 {
+				return errors.New("only car files version 1 supported at present")
+			}
+
+			for _, c := range car.Header.Roots {
+				roots[c] = struct{}{}
+			}
+
+			for {
+				block, err := car.Next()
+				if err != nil && err != io.EOF {
+					return err
+				} else if block == nil {
+					break
+				}
+
+				// the double-decode is suboptimal, but we need it for batching
+				nd, err := ipld.Decode(block)
+				if err != nil {
+					return err
+				}
+
+				if err := batch.Add(ctx, nd); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		}()
+
+		if err != nil {
+			ret <- importResult{err: err}
+			return
+		}
+	}
+
+	if err := it.Err(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	if err := batch.Commit(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	ret <- importResult{roots: roots}
+}
+
+// Export exports the DAG rooted at a given CID into a CAR file. The error channel will return at most one error and
+// the reader is usable before the error channel has been closed.
+func (api *dagAPI) Export(ctx context.Context, c cid.Cid) (io.ReadCloser, <-chan error) {
```

Comment:
I'm much happier with the `Export` API, although returning error channels still doesn't feel great here. Up for suggestions.

Reply:
Suggestion: Instead of returning an error channel, what about calling [`CloseWithError`](https://golang.org/pkg/io/#PipeWriter.CloseWithError) on pipeW.  That way the caller will get the error when trying to read from the `io.ReadCloser` returned here, if an error happens before reading is finished.
Example: https://play.golang.org/p/5caJVZb869-

Reply:
Thanks, that looks useful! @ribasushi do you recall any reason we didn't do this before (e.g. it didn't play well with the commands lib emitter)?

Reply:
I simply didn't know about `CloseWithError` at the time: 👍